File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
class Solution {
    
    // 숙련도에 따른 퍼즐 통과 시간
    private long passPuzzleTime (int[] diffs, int[] times, int level, long limit) {
        long totalTime = 0;
        for (int i = 0; i < diffs.length; i++) {
            int time_prev = (i == 0) ? 0 : times[i - 1];
            int time_cur = times[i];
            int levelDiff = diffs[i] - level;
            if (levelDiff >= 0) {
                totalTime += (long) (time_cur + time_prev) * levelDiff + time_cur;                
            } else {
                totalTime += (long) time_cur;
            }
            if (limit < totalTime) {
                return totalTime;
            }
        }
        return totalTime;
        
    }
    public int solution(int[] diffs, int[] times, long limit) {

        // 숙련
        int start = 1; 
        int level = 0;
        int end = 100000; 
        while (start <= end) {
            int midLevel = (start + end) / 2;
            long passTime = passPuzzleTime(diffs, times, midLevel, limit);
            if (passTime <= limit) {
                level = midLevel;
                // 30000
                // 1 ~ 29999 => 중간을 찾아서
                // 15000
                // 1 ~ 14999 => 중간 
                end = midLevel - 1;   
                // end = midLevel;
            } else {
                start = midLevel + 1;
                // start = midLevel;
            }   

        }
        return level;
    }
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.js
function solution(diffs, times, limit) {
    
    // 현재 퍼즐의 난이도 = diff
    // 현재 퍼즐의 소요 시간 = time_cur
    // 이전 퍼즐의 소요시간 = time_prev
    // 현재 숙련도 = level
    
    // 2차 풀이: level을 min, max의 중간값으로 찾기
    let min_diff = 1;
    let max_diff = 100_000;    
    let level = 0;
    const m = diffs.length;
             
    while (min_diff <= max_diff) {        
        const mid = Number.parseInt((min_diff + max_diff) / 2)         
        if (can_pass_puzzle(mid)) {
            level = mid
            max_diff = mid - 1
        } else {
            min_diff = mid + 1
        }                
    }
    
    function can_pass_puzzle(cur_level) {
        let cur_limit = limit - times[0]
        for (let i = 1; i < m; i++) {
            const cur_diff = diffs[i]
            cur_limit -= calc_time(cur_level, cur_diff, times[i], times[i - 1]);
            if (cur_limit < 0) {
                return false
            }
        }
        return cur_limit >= 0
    }
    
    return level;
}

function calc_time(cur_level, cur_diff, cur_time, prev_time) {
    if (cur_level >= cur_diff) {
        return cur_time
    }
    return (cur_time + prev_time) * (cur_diff - cur_level) + cur_time
}